### PR TITLE
feat: validate withdrawal storage root

### PR DESCRIPTION
**Pull request by chokobole**

### CODE REVIEW

File: components/node/sources/rollupclient.go [hunk 1]
```diff
@@ -18,9 +18,9 @@ func NewRollupClient(rpc client.RPC) *RollupClient {
 	return &RollupClient{rpc}
 }
 
-func (r *RollupClient) OutputAtBlock(ctx context.Context, blockNum uint64, includeNextBlock bool) (*eth.OutputResponse, error) {
+func (r *RollupClient) OutputAtBlock(ctx context.Context, blockNum uint64, isForFaultProof bool) (*eth.OutputResponse, error) {
```
Comment (Pangssu) [inline components/node/sources/rollupclient.go:21]: What about seperating them into `OutputAtBlock` and `OutputAtBlockWithProof` ?

File: packages/contracts/contracts/libraries/Types.sol [hunk 2]
```diff
@@ -189,4 +189,25 @@ library Types {
...
         bytes data;
     }
+
+    /**
+     * @notice Struct representing the data for verifying the public input.
+     *
+     * @custom:field srcOutputRootProof          Proof of the source output root.
+     * @custom:field dstOutputRootProof          Proof of the destination output root.
+     * @custom:field publicInput                 Ingredients to compute the public input used by ZK proof verification.
+     * @custom:field rlps                        Pre-encoded RLPs to compute the next block hash of the source output root proof.
```
Comment (Pangssu) [inline packages/contracts/contracts/libraries/Types.sol:199]: Is exceeding the max length allowed?

**Comment by seolaoh** [comment]
> When I run `TestERC20BridgeDeposits()` in `bridge_test.go`, it fails with an error of `"L2ToL1MessagePasser: storage layout entry for msgNonce not found"`. Could you check test?

You need to update bindings!

File: packages/contracts/contracts/libraries/Types.sol (hunk 2 above)
Comment (chokobole) [reply packages/contracts/contracts/libraries/Types.sol:199]: Is there a maximum length in solidity formatting rule?

File: components/node/sources/rollupclient.go (hunk 1 above)
Comment (chokobole) [reply components/node/sources/rollupclient.go:21]: What if `PublicInputProof` wraps all the field required for a public input and rename that flag to `includePublicInputProof`? Does it make sense?

```go
type OutputResponse struct {
	Version               Bytes32           `json:"version"`
	OutputRoot            Bytes32           `json:"outputRoot"`
	BlockRef              L2BlockRef        `json:"blockRef"`
	NextBlockRef          L2BlockRef        `json:"nextBlockRef"`
	WithdrawalStorageRoot common.Hash       `json:"withdrawalStorageRoot"`
	StateRoot             common.Hash       `json:"stateRoot"`
	Status                *SyncStatus       `json:"syncStatus"`
	PublicInputProof      *PublicInputProof `json:"publicInputProof"`
}
```

**Comment by chokobole** [comment]
@0xHansLee @seolaoh Sorry about that. On my local PC, I definitely ran `make more` and I don't know why this gives me an output whose storage is empty.

File: packages/contracts/deploy/009-Colosseum.ts [hunk 3]
```diff
@@ -19,6 +19,7 @@ const deployFn: DeployFunction = async (hre) => {
     hre,
     'SecurityCouncilProxy'
   )
+  const zkMerkleTrie = await getDeploymentAddress(hre, 'ZKMerkleTrie')
```
Comment (seolaoh) [inline packages/contracts/deploy/009-Colosseum.ts:22]:
```suggestion
  const zkMerkleTrieAddress = await getDeploymentAddress(hre, 'ZKMerkleTrie')
```

File: packages/contracts/contracts/L1/ZKTrieHasher.sol [hunk 4]
```diff
@@ -75,31 +75,65 @@ contract ZKTrieHasher {
...
+
+    /**
+     * @notice Computes a root hash of 2 elements.
+     *
+     * @param left_leaf  Bytes32 left leaf to be hashed.
+     * @param right_leaf Bytes32 right leaf to be hashed.
+     *
+     * @return A hash of 2 elements.
+     */
+    function _hashFixedElems2(bytes32 left_leaf, bytes32 right_leaf) internal view returns (bytes32) {
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/ZKTrieHasher.sol:121]: How about changing the function name to `_hashFixed2Elems`, `_hashFixed3Elems`? I think it's more understandable.

File: bindings/bindings/zkmerkletrie.go
Comment (seolaoh) [inline bindings/bindings/zkmerkletrie.go:1]: When I execute `make bindings`, this file does change.

File: packages/contracts/contracts/libraries/Types.sol (hunk 2 above)
Comment (Pangssu) [reply packages/contracts/contracts/libraries/Types.sol:199]: Yes, it is.
FYI, https://docs.soliditylang.org/en/latest/style-guide.html#maximum-line-length

File: components/node/sources/rollupclient.go (hunk 1 above)
Comment (Pangssu) [reply components/node/sources/rollupclient.go:21]: good

Comment (chokobole) [reply components/node/sources/rollupclient.go:21]: Sorry :( I don't have much time to resolve this requirement. Could somebody else do this for me..?

**Comment by chokobole** [comment]
@seolaoh could you check `make bindings` again?

File: packages/contracts/contracts/L1/ZKTrieHasher.sol (hunk 4 above)
Comment (chokobole) [reply packages/contracts/contracts/L1/ZKTrieHasher.sol]: renamed.

File: packages/contracts/deploy/009-Colosseum.ts (hunk 3 above)
Comment (chokobole) [reply packages/contracts/deploy/009-Colosseum.ts]: renamed.

File: packages/contracts/contracts/libraries/Types.sol (hunk 2 above)
Comment (chokobole) [reply packages/contracts/contracts/libraries/Types.sol]: I wrapped line.

**Review by seolaoh — APPROVED** [review]
LGTM

File: components/validator/challenger.go [hunk 5]
```diff
@@ -440,15 +440,59 @@ func (c *Challenger) GetChallenge(ctx context.Context, outputIndex *big.Int) (bi
...
+}
+
+func (c *Challenger) OutputWithProofAtBlockSafe(ctx context.Context, blockNumber uint64) (*eth.OutputResponse, error) {
+	ctx, cancel := context.WithTimeout(ctx, c.cfg.NetworkTimeout)
+	defer cancel()
+	return c.cfg.RollupClient.OutputWithProofAtBlock(ctx, blockNumber)
+}
+
+func (c *Challenger) PublicInputProof(ctx context.Context, blockNumber uint64) (bindings.TypesPublicInputProof, error) {
+	srcOutput, err := c.OutputAtBlockSafe(ctx, blockNumber, true)
```
Comment (0xHansLee) [inline components/validator/challenger.go:456]:
```suggestion
	srcOutput, err := c.OutputAtBlockSafe(ctx, blockNumber)
```

File: components/validator/challenger.go [hunk 6]
```diff
@@ -440,15 +440,59 @@ func (c *Challenger) GetChallenge(ctx context.Context, outputIndex *big.Int) (bi
...
 	if err != nil {
-		return nil, err
+		return bindings.TypesPublicInputProof{}, err
+	}
+
+	dstOutput, err := c.OutputAtBlockSafe(ctx, blockNumber+1, false)
```
Comment (0xHansLee) [inline components/validator/challenger.go:461]:
```suggestion
	dstOutput, err := c.OutputAtBlockSafe(ctx, blockNumber+1)
```

**Review by 0xHansLee — CHANGES_REQUESTED** [review]
The type of argument (`blockRef`) of `FetchProofAndPair` in `e2e/e2eutils/mock_proof_fetcher.go` should be changed to `uint64`

**Review by 0xHansLee — APPROVED** [review]
lgtm

**Review by Pangssu — APPROVED** [review]
lgtm

**Pull request merged**